### PR TITLE
Add safe exit

### PR DESCRIPTION
**Pull request by J02M**

### CODE REVIEW

**Comment by lechat** [comment]
If that test have to be placed last, then please place it last.

**Comment by J02M** [comment]
> If that test have to be placed last, then please place it last.

How could I do it, do you have a suggestion (my first python contribution)

**Comment by lechat** [comment]
I think that you need to add `pytest-ordering` to test-requirements.txt and then decorate test function with `@pytest.mark.last`

See more details here: https://pytest-ordering.readthedocs.io/en/develop/

**Comment by J02M** [comment]
Seems Travis CI is failing for 2.7 https://travis-ci.com/pycontribs/jenkinsapi/jobs/192151866, due to circumstances not related to my changes, what is the procedure? should i just force push again, and hope for the best?

File: test-requirements.txt [hunk 1]
```diff
@@ -1,6 +1,7 @@
 pytest==3.7.0
 pytest-mock
 pytest-cov
+pytest-ordering
```
Comment (lechat) [inline test-requirements.txt]: There is no need for this anymore. Please remove.

File: jenkinsapi/jenkins.py [hunk 2]
```diff
@@ -463,6 +463,24 @@ def _wait_for_reboot(self):
...
+                                                      valid=valid)
+        if wait_for_exit:
+            self._wait_for_exit()
+        return resp
+
+    def _wait_for_exit(self):
+        # We need to make sure all jobs have finished,
+        # and that jenkins is unavailable
+        # One way to be sure is to make sure jenkins is really down.
+        self.__jenkins_is_unavailable()  # Blocks until jenkins is unavailable
```
Comment (lechat) [inline jenkinsapi/jenkins.py]: I don't think that this code will block if Jenkins is available -  it will return immediately if Jenkins still replies with 200, 503 and 500. it is better to add actual check that will exit only on `ConnectionError`.

File: jenkinsapi_tests/systests/test_quiet_down.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+"""
+System tests for setting jenkins in quietDown mode
+"""
+
+
+def test_quiet_down(jenkins):
```
Comment (lechat) [inline jenkinsapi_tests/systests/test_quiet_down.py]: There are no test assertions. What is the point of this test?

Comment (J02M) [reply jenkinsapi_tests/systests/test_quiet_down.py]: The purpose is to make sure the added functionality gets executed during testing. Unfortunately, there is currently no support (as far as I have been able to see) in the code for requesting <server_url>/api/xml 
This would make it possible to check the value of:  <quietingDown>false</quietingDown>

Comment (J02M) [reply jenkinsapi_tests/systests/test_quiet_down.py]: I'can see Travis is failing everything, but I'm not able to figure out, what I can do to fix this. Is it caused by my changes?

File: jenkinsapi/jenkins.py [hunk 4]
```diff
@@ -22,13 +22,16 @@
 from jenkinsapi.nodes import Nodes
```
Comment (lechat) [inline jenkinsapi/jenkins.py]: Please revert this change. This is already done by parent class.

File: jenkinsapi_tests/systests/test_quiet_down.py [hunk 5]
```diff
@@ -1,11 +1,28 @@
...
 
-def test_quiet_down(jenkins):
+
+@pytest.mark.run_these_please
+def test_quiet_down_and_cancel_quiet_down(jenkins):
     jenkins.poll()  # jenkins should be alive
     jenkins.quiet_down()  # put Jenkins in quietDown mode
+
+    jenkins_api = jenkins.get_api_python_data()
+    is_quieting_down = jenkins_api['quietingDown']
```
Comment (lechat) [inline jenkinsapi_tests/systests/test_quiet_down.py]: This code is not part of test. There shall be a property or method in `Jenkins` object that returns `quietingDown` value.

File: jenkinsapi/jenkins.py (hunk 4 above)
Comment (J02M) [reply jenkinsapi/jenkins.py]: I'll be happy to remove it, but I can't recall having added the line!

Comment (lechat) [reply jenkinsapi/jenkins.py]: I mean whole change made in commit 50f17a0 to `jenkins.py` is not needed and must be reverted.

Comment (J02M) [reply jenkinsapi/jenkins.py]: is there a better way to request the current state of api/python?

Comment (J02M) [reply jenkinsapi/jenkins.py]: do you still want me to remove: "from jenkinsapi.nodes import Nodes"
This i not a added by me?

File: jenkinsapi_tests/systests/test_quiet_down.py (hunk 5 above)
Comment (J02M) [reply jenkinsapi_tests/systests/test_quiet_down.py]: Would it be possible to implement getting the 'api/python' in jenkinsbase.py, parsing the content to a python object, and return the python object. This would make it a one-time endeavour, making it compatible with any content in this object?

**Comment by J02M** [comment]
Seems the test setup is erratic, is there some way I can contribute to reducing this?

**Comment by lechat** [comment]
It is erratic because it is either can't download plugins (jenkins.io was down?) or jenkins just decides to not to respond at random.

**Comment by J02M** [comment]
consider downloading plugins from Jenkins artifactory (https://repo.jenkins-ci.org), which is much more stable!

**Comment by J02M** [comment]
Please review merge request.

**Comment by J02M** [comment]
Any review progress?

**Comment by J02M** [comment]
Everything is green again after the last rebase, please review.

**Comment by J02M** [comment]
Thx for your collaboration it has been a pleasure.
